### DISCUSSION TITLE
How to place the marker value fixed on top of each stacked bar

### DISCUSSION THREAD

**Discussion by virtualpathum** [patrykandpatrick/vico#589] [Ideas]
Hi, I want to achieve the following. It is a simple chart and just wanted to show the Y-axis value on top of each bar. How can I achieve this?
![chart](https://github.com/patrykandpatrick/vico/assets/3455658/831d6cec-e432-4308-a81f-6b702a421029)

**Comment by patrickmichalik** [comment]
Hello! I’m afraid this isn’t possible yet. Vico does support data labels (see the `dataLabel` parameters), but in such a case, there’d be a shared data label above each column stack. We’ll look into adding the feature in question—I’m moving this to the “Ideas” category.

**Reply by virtualpathum** [reply]
Hi, Thanks for the quick reply. We plan to go live within the next two months and the chart will be showing on a TV screen rather than a device the user can touch. So, we need to show the values in each bar. Is there a possibility of giving us a patch version before you add this feature?

I tried the dataLabel but couldn't figure out how to use it for this purpose.

btw thanks again for the prompt replies.

**Reply by Gowsky** [reply]
With all due respect, asking FOSS maintainers to do your job and spend their very limited time on some special request, rather than more widely needed or requested changes, is a bit out of place. If you need such functionality before we implement and release it, you can fork the repository and make the needed changes yourself. Then you may also consider opening a pull request, so others can benefit from it too.

If you want to avoid forking, you can try implementing this as a custom `Decoration`. The API lets you draw anything you’d like, and its functions contain information needed to calculate the labels' positions. You can use `TextComponent` to draw the text.

**Reply by virtualpathum** [reply]
@Gowsky Yeah.. I understand that making such a request is unreasonable. However, we have now changed our requirement to follow the current marker design in the lib with some minor customizations. Still, doing such minor customization is also taking so much time.  I know that is because the library has a lot of capabilities and we need to understand it fully/largely before doing anything.

If you guys could provide some more generic examples to show how each major component works and where to change, that would be useful for everybody. For eg: I couldn't find any marker customization in the given sample charts...

Anyway, keep up your good work and thanks for your prompt replies.

**Reply by Gowsky** [reply]
Adding various `Marker` implementations in the sample app is on our to-do list. I want to highlight that the `rememberMarker` function from the sample app is an example of `Marker` customization, as it creates a custom `Marker` via `MarkerComponent`. It’s not a default included in the library but rather an example of how to use the library’s APIs. You can use the function and edit it to suit your requirements. I also want to note that [`Marker`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.10/core/com.patrykandpatrick.vico.core.marker/-marker/) is an interface that you can implement however you like. You can make direct calls to the `Canvas` or use existing Vico `Component`s like `ShapeComponent` or `TextComponent` that make drawing onto the `Canvas` easier. The `MarkerComponent` is the default implementation of the interface. By looking at its source code you can learn how certain `Component`s are being positioned so that you can write your implementation more easily.

If you have any specific questions, please let me know.

**Reply by virtualpathum** [reply]
Hi @Gowsky @patrickmichalik  Thanks. My requirement is this
![image](https://github.com/patrykandpatrick/vico/assets/3455658/2e66d7c6-71a6-40d8-8d11-780b8edc1919)

So far, as I understand, to add a prefix to each `model.entry.y` that prefix value should also be an element of 

```
public data class EntryModel(
        public val location: Point,
        public val entry: CartesianLayerModel.Entry,
        public val color: Int,
        public val index: Int,
 )
```

this data class in the Marker Interface. I am not sure about the impact of adding another value here. Therefore, I did this **dirty fix** for the time being to my `CustomMarkerLabelFormatter`
 
```
class CustomMarkerLabelFormatter(private val colorCode: Boolean = true, private val topBarLabel: String,  private val bottomBarLabel: String) : MarkerLabelFormatter {
    override fun getLabel(
        markedEntries: List<Marker.EntryModel>,
        chartValues: ChartValues,
    ): CharSequence =
        markedEntries.transformToSpannable(
            prefix = if (markedEntries.size > 1) PATTERN.format(markedEntries.sumOf { it.entry.y }.toInt()) + "" else "",
            postfix = if (markedEntries.size > 1) " " else "",
            separator = " ",
        ) { model ->
            if (colorCode) {
                val prefix = if(model.color == Color(0xFF2196F3).toArgb()){
                    topBarLabel
                } else {
                    bottomBarLabel
                }
                appendCompat(
                    prefix,
                    PATTERN.format(model.entry.y.toInt()),
                    ForegroundColorSpan(model.color),
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE,
                )
            } else {
                append(PATTERN.format(model.entry.y))
            }
        }

    private val CartesianLayerModel.Entry.y
        get() =
            when (this) {
                is ColumnCartesianLayerModel.Entry -> y
                is LineCartesianLayerModel.Entry -> y
                else -> throw IllegalArgumentException("Unexpected `CartesianLayerModel.Entry` implementation.")
            }

    private companion object {
        const val PATTERN = "%d"
    }
}
```
Appreciate it if you could suggest a better way to do this. Thanks

**Reply by virtualpathum** [reply]
Just found out that ` MarkerComponent.LabelPosition = MarkerComponent.LabelPosition.AroundPoint/AbovePoint` doesn't move the marker to the top of the stacked bars as I want, Is there any trick to do that?

**Reply by patrickmichalik** [reply]
@virtualpathum, regarding label formatting, `Marker.EntryModel` instances are created internally, and the class should be left as is. All you have to do is override `MarkerLabelFormatter#getLabel` as required. Rather than comparing colors, however, you should consider the index of a given `Marker.EntryModel` in `markedEntries` (not to be confused with `Marker.EntryModel#index`). The first item in the list corresponds to the bottom column, and the second one corresponds to the top column (much like the first series corresponds to the bottom columns, and the second one corresponds to the top columns). I see that you’ve copied over the `transformToSpannable` function. You can update it such that the lambda has an index parameter. That said, `transformToSpannable` doesn’t have to be used—you can create a `SpannableString` more directly. `DefaultMarkerLabel` has to handle all possible setups, but your custom `MarkerLabelFormatter` can be significantly simpler. This is most likely sufficient:

```kt
SpannableStringBuilder("${markedEntries.sumOf { it.entry.y }.toInt()}")
    .appendCompat(
        " $topBarLabel${markedEntries[0].entry.y.toInt()}",
        ForegroundColorSpan(markedEntries[0].color),
        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE,
    )
    .appendCompat(
        " $bottomBarLabel${markedEntries[1].entry.y.toInt()}",
        ForegroundColorSpan(markedEntries[1].color),
        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE,
    )
```

As for `MarkerComponent.LabelPosition.AbovePoint` and `MarkerComponent.LabelPosition.AroundPoint`, it looks like there’s a positioning bug. We’ve fixed it and will release an update soon. For the time being, you can override `MarkerComponent#drawLabel` (`MarkerComponent` is an open class) and apply the change from [this commit](https://github.com/patrykandpatrick/vico/commit/782c9337ba06fa8efee81f5fe5a0db37bb8b7a61).

**Reply by patrickmichalik** [reply]
[Vico 2.0.0 Alpha 11](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.11) resolves the `MarkerComponent.LabelPosition` issue.

**Reply by virtualpathum** [reply]
@patrickmichalik thanks for the update. These solutions fixed my chart issues!

**Reply by patrickmichalik** [reply]
You’re welcome, @virtualpathum! I’m glad to hear that.
